### ISSUE TITLE
Expose Provisioner's Knowledge to Plugins

### ISSUE THREAD

**Issue by boumenot** [mitchellh/packer#3170]
I logged an [issue](https://github.com/Azure/packer-azure/issues/200) against packer-azure, but I wanted to raise it here.  I may be misunderstanding the way it works, or I may not have a good enough understanding so I am looking for guidance.

Provisioner's are configured by the user to execute remote commands, and packer-azure would like to take advantage of this so it can execute remote commands too.  For example, the Azure plugin executes a generalize OS step, and does so by executing a remote command.  The way remote commands are executed differs between VM images.  Some can just execute the command as is, and others need to inject the root password via STDIN to ensure sudo properly elevates.  The knowledge of how this is done is perfectly captured in provisioners, but I don't see how that functionality is exposed.  There's a provisioner hook, but I don't see how to update the configuration (call Prepare) with a new configuration for the script(s) that I want to execute.

I can solve the issue by having the user add the generalize command to his provisioner configuration, but I would prefer they not have to worry about it.  If they choose not to execute the command, or accidentally miss it I would like to issue a warning; I did not see how the provisioner configuration is exposed to plugins to do this.

**Comment by cbednarski** [comment]
> I did not see how the provisioner configuration is exposed to plugins to do this.

The short answer is it isn't. The provisioner _is_ a plugin and the scope of its configuration is largely isolated to that plugin.

Currently, the communicator is configured and then passed into the provisioner. This means that the provisioner does not know how to connect to the machine, but _does_ know how to execute commands once it is connected. However, this assumes three things: 1. That the means to connect over SSH (for example) already exists. 2. That the communicator does not need to change. 3. That one set of credentials is sufficient for doing all of the remote provisioning.

To make the assumption 1 happen, user information is usually setup in the guest OS by a kickstart file or other unattended install file that is provided to the VM over direct console access (e.g. VNC, where we simply pass keystrokes to the VM as if we were sitting in front of it). For instance this might set the root password on the VM and then create a basic user to login with.

Assumption 2 potentially causes problems. Because the communicator is configured at the builder level and not the provisioner, different provisioners for the same build are stuck using the same communicator. Because of the pattern in assumption 1, above, this is usually OK. This is also mitigated by assumption 3, wherein we expect that the credentials provisioned in 1 will satisfy all cases for 2.

So to recap, a builder has one communicator which assumes that the credentials already exist, and that one communicator is used for all provisioning. If you wanted to use SSH _and_ WinRM in the same build, we don't support that right now.

**Comment by boumenot** [comment]
I'm looking for a solution to 2.  A provisioner knows how to _format_ a command for a communicator, and in my case it would be valuable because different images require different formats.

**Comment by cbednarski** [comment]
Sorry to keep you waiting on this one! What you're describing is actually the reason why communicators and provisioners are separated.

Communicators communicate over a protocol (SSH, WinRM, SFTP, etc.) but don't know what type of OS they are talking to. The provisioner does.

Let's take some examples:

- SSH communicating with a Windows guest. In this case, how do we execute things? We can't treat the guest like linux, obviously.
- SSH communicating with a Linux guest using a root user. In this case you can do lots of things.
- SSH communicating with a Linux guest using a non-root user. In this case you need to sudo to run certain commands.
- SSH communicating with a FreeBSD guest using tcsh instead of bash. In this case environment variables need to be treated differently.

So in each of these cases you need to know something about the guest OS. The communicator doesn't know this. In fact, it _can't_ know this. SSH is a only protocol. The provisioner does have this info, however, so it can do the right thing.

> For example, the Azure plugin executes a generalize OS step, and does so by executing a remote command. ... I can solve the issue by having the user add the generalize command to his provisioner configuration, but I would prefer they not have to worry about it.

I guess I don't understand what you're trying to do here. It seems like you would need to use a provisioner to do this. If it's something that's unique to Azure but affects all kinds of guests, you could make an Azure provisioner that negotiates the right type of shell on the remote side automatically. OTOH an example in the docs using the SSH / WinRM provisioners might suffice too.

**Comment by boumenot** [comment]
I've decided to just embedded the command in the provisioner.  If we ever have a chance to chat again I'll try to explain myself better. :smile:   This is minor and, I have a more than acceptable work around, so I'm going to close the issue.
